src/main/java/com/atlassian/db/replica/internal/ReplicaConnectionProvider.java

```diff
@@ -1,104 +1,75 @@
-            return connectionProvider.getMainConnection();
-        }
-    };
+    private final Warnings warnings;
 
     public ReplicaConnectionProvider(
         ConnectionProvider connectionProvider,
-        ReplicaConsistency consistency
+        ReplicaConsistency consistency,
+        StateListener stateListener
     ) {
-        this.connectionProvider = connectionProvider;
+        this.parameters = new ConnectionParameters();
+        this.warnings = new Warnings();
+        this.state = new ConnectionState(connectionProvider, consistency, parameters, warnings, stateListener);
         this.consistency = consistency;
     }
 
-    private void initialize(Connection connection) throws SQLException {
-        if (!initializedConnections.contains(connection)) {
-            if (isAutoCommit != null) {
-                connection.setAutoCommit(isAutoCommit);
-            }
-            if (transactionIsolation != null) {
-                connection.setTransactionIsolation(transactionIsolation);
-            }
-            if (catalog != null) {
-                connection.setCatalog(catalog);
-            }
-            if (typeMap != null) {
-                connection.setTypeMap(typeMap);
-            }
-            if (holdability != null) {
-                connection.setHoldability(holdability);
-            }
-            initializedConnections.add(connection);
-        }
+    public Connection getWriteConnection() throws SQLException {
+        return state.getWriteConnection();
+    }
+
+    public Connection getReadConnection() throws SQLException {
+        return state.getReadConnection();
     }
 
     public void setTransactionIsolation(Integer transactionIsolation) {
-        this.transactionIsolation = transactionIsolation;
-        initializedConnections.clear();
+        parameters.setTransactionIsolation(transactionIsolation);
     }
 
     public int getTransactionIsolation() throws SQLException {
-        if (this.transactionIsolation != null) {
-            return this.transactionIsolation;
+        if (parameters.getTransactionIsolation() != null) {
+            return parameters.getTransactionIsolation();
         } else {
-            return getWriteConnection().getTransactionIsolation();
+            return state.getWriteConnection().getTransactionIsolation();
         }
     }
 
-    public void setAutoCommit(Boolean autoCommit) {
-        final Boolean autoCommitBefore = getAutoCommit();
-        this.isAutoCommit = autoCommit;
-        initializedConnections.clear();
-        if (!autoCommitBefore.equals(getAutoCommit())) {
+    public void setAutoCommit(Boolean autoCommit) throws SQLException {
+        final boolean autoCommitBefore = getAutoCommit();
+        parameters.setAutoCommit(autoCommit);
+        final Optional<Connection> connection = state.getConnection();
+        if (connection.isPresent()) {
+            connection.get().setAutoCommit(autoCommit);
```

Comment:
Can't use `Optional.ifPresent` due to `throws SQLException`
Checked exceptions trying to help again 😂